smslibrary/src/main/java/com/eis/smslibrary/SMSReceivedBroadcastReceiver.java

```diff
@@ -0,0 +1,63 @@
+package com.eis.smslibrary;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+import android.os.Bundle;
+import android.telephony.SmsMessage;
+
+/**
+ * Broadcast receiver for received messages, called by Android.
+ *
+ * @author Luca Crema, Marco Mariotto
+ * @since 29/11/2019
+ *
+ */
+public class SMSReceivedBroadcastReceiver extends BroadcastReceiver {
+
+    /**
+     * Parses message and calls listener
+     */
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        Bundle extras = intent.getExtras();
+        if (extras == null) {
+            return;
+        }
+
+        Object[] smsExtras = (Object[]) extras.get("pdus");
+        String format = (String) extras.get("format");
+
+        if(smsExtras == null) //could be null
+            return;
+
+        for (Object smsData : smsExtras) {
+            SmsMessage receivedSMS = createMessageFromPdu(smsData, format);
+            String smsContent = receivedSMS.getMessageBody();
+            String phoneNumber = receivedSMS.getOriginatingAddress();
+
+            if (phoneNumber == null) //could be null
+                return;
+
+            SMSMessage parsedMessage = SMSMessageHandler.getInstance().parseMessage(phoneNumber, smsContent);
+            if (parsedMessage != null) {
+                SMSHandler.getInstance().callReceivedListener(parsedMessage);
+            }
+        }
+    }
+
+    /**
+     * Calls the appropriate method to create a message from its pdus
+     * @param smsData message pdus
+     * @param format available only on build version >= 23
+     * @return
+     */
+    private SmsMessage createMessageFromPdu(Object smsData, String format) {
```

Comment:
A method exists in Android libraries to get an SmsMessage array from an Intent without having to go through all the pain: Telephony.Sms.Intents.getMessagesFromIntent(intent)

Reply:
Requires API 19, current one is 15

Reply:
I see then it's ok, we used 21 and therefore didn't get an error about it